pym/bob/input.py

```diff
@@ -603,6 +603,72 @@ def isDeterministic(self):
     def hasJenkinsPlugin(self):
         return True
 
+
+class CvsScm(BaseScm):
+    # Checkout using CVS
+    # - mandatory parameters: cvsroot, module
+    # - optional parameters: rev, dir (dir is required if there are multiple checkouts)
+    def __init__(self, spec):
+        super().__init__(spec)
+        self.__cvsroot = spec["cvsroot"]
+        self.__module = spec["module"]
+        self.__rev = spec.get("rev")
+        self.__dir = spec.get("dir", ".")
+
+    def resolveEnv(self, env):
+        # Resolve macros in CVS names
+        super().resolveEnv(env)
+        self.__cvsroot = env.substitute(self.__cvsroot, "cvs::cvsroot")
+        self.__module = env.substitute(self.__module, "cvs::module")
+        if self.__rev:
+            self.__rev = env.substitute(self.__rev, "cvs::rev")
+        if self.__dir:
+            self.__dir = env.substitute(self.__dir, "cvs::dir")
+
+    def asScript(self):
+        revarg = "-r {rev}".format(rev=self.__rev) if self.__rev != None else "-A"
+        return """
+# Validate that we're talking to the correct repository. If not, ditch it.
```

Comment:
AFAICT this check is unnecessary. Bob will always check the digests that are returned by getDirectories() an will move reconfigured directories to an "attic" directory next to the workspace. The idea is that switching branches/tags in a working copy is hard to implement for any SCM. The safe choice is to move away offending directories and make a clean checkout with the new configuration. In release mode (bob build ...) Bob will always use a fresh directory anyway.

Therefore the checkout script may ignore any changes made to the recipe. The only thing that should be handled gracefully are manual changes in the working copy by the user without modifying the recipe.